binary-oracle-pair/program/src/processor.rs

```diff
@@ -284,10 +331,183 @@ impl Processor {
             authority_account_info.clone(),
             amount,
         )?;
-        
+
+        Ok(())
+    }
+
+    /// Process Withdraw instruction
+    pub fn process_withdraw(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        amount: u64,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        let pool_account_info = next_account_info(account_info_iter)?;
+        let authority_account_info = next_account_info(account_info_iter)?;
+        let pool_deposit_token_account_info = next_account_info(account_info_iter)?;
+        let token_pass_user_account_info = next_account_info(account_info_iter)?;
+        let token_fail_user_account_info = next_account_info(account_info_iter)?;
+        let token_pass_mint_info = next_account_info(account_info_iter)?;
+        let token_fail_mint_info = next_account_info(account_info_iter)?;
+        let user_token_destination_account_info = next_account_info(account_info_iter)?;
+        let token_program_id_info = next_account_info(account_info_iter)?;
+
+        if amount == 0 {
+            return Err(PoolError::InvalidAmount.into());
+        }
+
+        let user_pass_token_account = Account::unpack(&token_pass_user_account_info.data.borrow())?;
+        let user_fail_token_account = Account::unpack(&token_fail_user_account_info.data.borrow())?;
+        let user_token_destination_account =
+            Account::unpack(&user_token_destination_account_info.data.borrow())?;
+
+        if user_pass_token_account.owner != user_fail_token_account.owner
+            && user_pass_token_account.owner != user_token_destination_account.owner
+        {
+            return Err(PoolError::InvalidAccountsOwner.into());
+        }
+
+        let pool = Pool::unpack(&pool_account_info.data.borrow())?;
+
+        let authority_pub_key =
+            Self::authority_id(program_id, pool_account_info.key, pool.bump_seed)?;
+        if *authority_account_info.key != authority_pub_key {
+            return Err(PoolError::InvalidAuthorityAccount.into());
+        }
+
+        if pool.decision == None {
```

Comment:
Better to use `if let` for `Some` and unwrapping decision flag at the same time, then on `else` process `None`

Reply:
Also, burn both tokens only if current slot is less than `mint_end_slot` or more than `decide_end_slot` and no decision was made. If slot is in the decision making interval and decision was not made yet — then throw exception.